### ISSUE TITLE
incomplete `scroll` paints when using opengl:gtk

### ISSUE THREAD

**Issue by aerusso** [Xpra-org/xpra#4201]
**Describe the bug**
With opengl GTK sometimes remote windows appear "clear".

**To Reproduce**
1. server command: `xpra start :$display --start-new-commands=yes --daemon=no --systemd-run=no`
2. client command: `xpra attach ssh://$user@$host/$display --opengl=yes:gtk -dpaint`
3. this was the nextcloud client app, I clicked on the drop-down menu for the user, repeatedly.  I was also able to get similar artifacts when using the "boats" program from `http://boats.sf.net` (comes with Debian).  (I have no clue what this program/game does, I just know that it depends on `libqt5gui5t64` on Debian, and I suspect that it's exercising a similar path as nextcloud-desktop).  This program can create similar artifacts when you select the red "create mark" button, causing the cursor have an associated circled number that moves with the mouse when the mouse moves; just kind of move the mouse all around the canvas area, eventually the paints get confused.  It looks like programs that directly draw on a canvas are having problems (but I do not have a minimal reproducer program).

**System Information (please complete the following information):**
 - Server OS: Debian unstable
 - Client OS: Debian unstable
 - Xpra Server Version 6.0
 - Xpra Client Version 6.0 (REGRESSION relative to 5.0.8/cannot reproduce with 5.0.8)

**Additional context**
Here is the out put of the log with -d paint:

<details>
<summary> client log </summary>

```
2024-04-27 00:08:03,902 draw_region(0, 0, 447, 601, scroll, 11 bytes, 0, typedict({'flush': 1, 'window-size': (447, 601)}), [<function WindowClient._do_draw.<locals>.record_decode_time at 0x7f363b7459e0>])
2024-04-27 00:08:03,903 draw_region(0, 69, 447, 186, webp, 3496 bytes, 0, typedict({'rgb_format': 'BGRA', 'quality': 100, 'has_alpha': True, 'flush': 0, 'window-size': (447, 601)}), [<function WindowClient._do_draw.<locals>.record_decode_time at 0x7f363b745bc0>, <bound method ClientWindowBase.after_draw_refresh of GLClientWindow(2 : GLAreaBacking(0x2, (447, 601)))>])
2024-04-27 00:08:03,904 do_scroll_paints(<__gi__.GdkWaylandGLContext object at 0x7f365d7ba240 (GdkWaylandGLContext at 0x4b7e280)>, ((0, 266, 447, 81, 0, -1), (0, 266, 447, 81, 0, -2), (0, 266, 447, 81, 0, -3), (0, 266, 447, 81, 0, -4), (0, 266, 447, 81, 0, -5), (0, 266, 447, 81, 0, -6), (0, 266, 447, 81, 0, -7), (0, 266, 447, 81, 0, -8), (0, 266, 447, 81, 0, -9), (0, 266, 447, 81, 0, -10), (0, 266, 447, 81, 0, -11)), 1)
2024-04-27 00:08:03,904 copy_fbo(447, 601, 0, 0, 0, 0)
2024-04-27 00:08:03,904 swap_fbos()
2024-04-27 00:08:03,904 record_decode_time(True, ) wid=2, scroll: 447x601, 2.2ms
2024-04-27 00:08:03,904 GLAreaBacking(0x2, (447, 601)).gl_paint_rgb(BGRA, 332568 bytes, x=0, y=69, width=447, height=186, rowstride=1788, options=typedict({'rgb_format': 'BGRA', 'quality': 100, 'has_alpha': True, 'flush': 0, 'window-size': (447, 601), 'encoding': 'webp'}))
2024-04-27 00:08:03,905 gravity_adjust(0, 69, typedict({'rgb_format': 'BGRA', 'quality': 100, 'has_alpha': True, 'flush': 0, 'window-size': (447, 601), 'encoding': 'webp'})) window_size=(447, 601), size=(447, 601), gravity=10
2024-04-27 00:08:03,905 record_decode_time(True, ) wid=2, webp: 447x186, 1.6ms
2024-04-27 00:08:03,905 after_draw_refresh(True, '') pending_refresh=[], backing=GLAreaBacking(0x2, (447, 601))
2024-04-27 00:08:03,905 repaint(0, 0, 447, 601) widget=<Gtk.GLArea object at 0x7f3654d10400 (GtkGLArea at 0x4acd990)>
2024-04-27 00:08:03,905 draw_fbo(<cairo.Context object at 0x7f363b713670>)
2024-04-27 00:08:03,905 on_render(<Gtk.GLArea object at 0x7f3654d10400 (GtkGLArea at 0x4acd990)>, <__gi__.GdkWaylandGLContext object at 0x7f365d7ba240 (GdkWaylandGLContext at 0x4b7e280)>)
2024-04-27 00:08:03,906 do_present_fbo(<__gi__.GdkWaylandGLContext object at 0x7f365d7ba240 (GdkWaylandGLContext at 0x4b7e280)>) will blit [(0, 0, 447, 601)]
2024-04-27 00:08:03,906 window viewport for self.render_size=(447, 601) and (0, 0, 0, 0) with scale factor 1: (0, 0, 447, 601)
2024-04-27 00:08:03,906 do_gl_show(0)
2024-04-27 00:08:03,906 GLAreaBacking(0x2, (447, 601)).do_present_fbo() done
2024-04-27 00:08:03,931 draw_fbo(<cairo.Context object at 0x7f363b713670>)
```
</details>

<details>
<summary> server log </summary>

```
2024-04-27 00:08:03,917 client   3 @46.678 draw_region(0, 0, 447, 601, scroll, 11 bytes, 0, typedict({'flush': 1, 'window-size': (447, 601)}), [<function WindowClient._do_draw.<locals>.record_decode_time at 0x7f363b7459e0>])                                           
2024-04-27 00:08:03,917 client   3 @46.679 draw_region(0, 69, 447, 186, webp, 3496 bytes, 0, typedict({'rgb_format': 'BGRA', 'quality': 100, 'has_alpha': True, 'flush': 0, 'window-size': (447, 601)}), [<function WindowClient._do_draw.<locals>.record_decode_time at 0x7f363b745bc0>, <bound method ClientWindowBase.after_draw_refresh of GLClientWindow(2 : GLAreaBacking(0x2, (447, 601)))>])
2024-04-27 00:08:03,918 client   3 @46.679 do_scroll_paints(<__gi__.GdkWaylandGLContext object at 0x7f365d7ba240 (GdkWaylandGLContext at 0x4b7e280)>, ((0, 266, 447, 81, 0, -1), (0, 266, 447, 81, 0, -2), (0, 266, 447, 81, 0, -3), (0, 266, 447, 81, 0, -4), (0, 266, 447, 81, 0, -5), (0, 266, 447, 81, 0, -6), (0, 266, 447, 81, 0, -7), (0, 266, 447, 81, 0, -8), (0, 266, 447, 81, 0, -9), (0, 266, 447, 81, 0, -10), (0, 266, 447, 81, 0, -11)), 1)
2024-04-27 00:08:03,918 client   3 @46.679 copy_fbo(447, 601, 0, 0, 0, 0)                                                                                                                                                                                                                                                                                                                      
2024-04-27 00:08:03,918 client   3 @46.680 swap_fbos()                                                                                                                                                                                                                     
2024-04-27 00:08:03,918 client   3 @46.680 record_decode_time(True, ) wid=2, scroll: 447x601, 2.2ms                                                                            
2024-04-27 00:08:03,919 client   3 @46.680 GLAreaBacking(0x2, (447, 601)).gl_paint_rgb(BGRA, 332568 bytes, x=0, y=69, width=447, height=186, rowstride=1788, options=typedict({'rgb_format': 'BGRA', 'quality': 100, 'has_alpha': True, 'flush': 0, 'window-size': (447, 601), 'encoding': 'webp'}))                                                                                                                                                      
2024-04-27 00:08:03,919 client   3 @46.680 gravity_adjust(0, 69, typedict({'rgb_format': 'BGRA', 'quality': 100, 'has_alpha': True, 'flush': 0, 'window-size': (447, 601), 'encoding': 'webp'})) window_size=(447, 601), size=(447, 601), gravity=10                                                                                                                                                
2024-04-27 00:08:03,919 client   3 @46.680 record_decode_time(True, ) wid=2, webp: 447x186, 1.6ms                                                                                                                                                                                                    
2024-04-27 00:08:03,919 client   3 @46.680 after_draw_refresh(True, '') pending_refresh=[], backing=GLAreaBacking(0x2, (447, 601))                                                                                                                                                                   
2024-04-27 00:08:03,919 client   3 @46.680 repaint(0, 0, 447, 601) widget=<Gtk.GLArea object at 0x7f3654d10400 (GtkGLArea at 0x4acd990)>                                                                                                                
2024-04-27 00:08:03,919 client   3 @46.681 draw_fbo(<cairo.Context object at 0x7f363b713670>)                                                                                  
2024-04-27 00:08:03,921 client   3 @46.681 on_render(<Gtk.GLArea object at 0x7f3654d10400 (GtkGLArea at 0x4acd990)>, <__gi__.GdkWaylandGLContext object at 0x7f365d7ba240 (GdkWaylandGLContext at 0x4b7e280)>)               
2024-04-27 00:08:03,921 client   3 @46.681 do_present_fbo(<__gi__.GdkWaylandGLContext object at 0x7f365d7ba240 (GdkWaylandGLContext at 0x4b7e280)>) will blit [(0, 0, 447, 601)]                                                                                                                                                                                                                    
2024-04-27 00:08:03,921 client   3 @46.681 window viewport for self.render_size=(447, 601) and (0, 0, 0, 0) with scale factor 1: (0, 0, 447, 601)                                                                                                                                                    
2024-04-27 00:08:03,921 client   3 @46.681 do_gl_show(0)                                                                                                                                                                                                
2024-04-27 00:08:03,921 client   3 @46.681 GLAreaBacking(0x2, (447, 601)).do_present_fbo() done                                                                                                                  
2024-04-27 00:08:03,945 client   3 @46.706 draw_fbo(<cairo.Context object at 0x7f363b713670>)
```
</details>

I did trim the log from before the distortion began; I can try to get a more complete log if it would be helpful.

**Comment by totaam** [comment]
Can you include a screenshot of the problematic paints?
What GPU and driver are you using?
Can you reproduce it with `--opengl=force:native` ?
What about turning off scroll encoding? `--encodings=all,-scroll`

**Comment by aerusso** [comment]
1. ![weird-artifact](https://github.com/Xpra-org/xpra/assets/6502699/5fa20eef-b158-45dc-8933-4e9c2102a30a)

My desktop background is black, I've redacted some text with 255,0,0 red (i.e., there is no artifact on the titlebar).

The black background is "clear". I.e., if I drag that xpra client window over another window, I can see the other window, beneath it **through** the remote xpra window:

![weird-artifact-2](https://github.com/Xpra-org/xpra/assets/6502699/22b7dca1-badc-4ce6-91d8-9a05af386782)


That image was made by moving the kolourpaint window beneath the nextcloud window AFTER the corruption started.

The corruption often goes away after resizing or presumably any client repaint of the affected area.

2.

```
GLU=1.3
array-handlers=<OpenGL.arrays.numpymodule.NumpyHandler object at 0x7fd54cb51110>
backend=egl
max-viewport-dims=16384, 16384
opengl=4.6
platform=linux
pyopengl=3.1.7
renderer=AMD Radeon Graphics (radeonsi, gfx1103_r1, LLVM 17.0.6, DRM 3.57, 6.7.9-amd64)
safe=True
shading-language-version=4.60
success=True
texture-size-limit=16384
vendor=AMD
zerocopy=
```
3. I cannot reproduce with `--opengl=force:native`, presumably because it falls back to no acceleration:

```
2024-04-27 06:47:56,870 No OpenGL_accelerate module loaded: No module named 'OpenGL_accelerate'
2024-04-27 06:47:56,920 Warning: cannot import OpenGL window module native
2024-04-27 06:47:56,920  'EGLPlatform' object has no attribute 'GLX'
2024-04-27 06:47:56,920 Warning: no OpenGL backend module found
```

(I do not have pyopengl_accelerate installed).

4. I'm not able to reproduce with the client starting with `--encodings=all,-scroll`

**Comment by totaam** [comment]
I can reproduce it, and I have no idea how to fix it.
Although the context is now EGL / core context, the `do_scroll_paints` method is unchanged.
This is the one from v5.x:
https://github.com/Xpra-org/xpra/blob/85ee52e738a75ab63749c0b986bf7702cf135c5d/xpra/client/gl/gl_window_backing_base.py#L646-L716
And the current one:
https://github.com/Xpra-org/xpra/blob/c013b14a74a37c048e5bd848c4587dfd23912533/xpra/client/gl/backing.py#L596-L665

**Comment by totaam** [comment]
> `'EGLPlatform' object has no attribute 'GLX'`

Ah, is it a Wayland desktop? (then `native` isn't available)

**Comment by totaam** [comment]
So, this is a strange one.
I first thought that this was a true regression, but I can reproduce _some_ corruption with v5.x LTS, just less often and far less drastic.
It would seem that the new OpenGL paint code is to blame - despite being unchanged since v5.x
But I can also reproduce some corruption with `--opengl=no`.

If I don't make progress, I could disable `scroll` for the next update as this will "fix" things - at least temporarily, buying some time.

**Comment by bugreporter42** [comment]
opengl=force:native has the same issue (client version 6.0). I'm getting a lot of corruption between different windows (using shadow for a remote desktop).

Client version 4.4.5 works correctly.

**Comment by cherio** [comment]
Same issue when connecting from xpra 6.0-1 client (Arch linux, XFCE/GTK) to either another 6.0-1 server (Arch linux) or v6.0-r0 server (ubuntu 22.04) from xpra stable repository.

```
2024-05-13 16:14:27,738 Xpra GTK3 X11 client version 6.0-r0
2024-05-13 16:14:27,742  running on Linux 6.8.9-arch1-2
2024-05-13 16:14:27,742  cpython 3.12
2024-05-13 16:14:27,742  window manager is 'Xfwm4'
2024-05-13 16:14:27,941 GStreamer version 1.24.3
2024-05-13 16:14:27,968 created unix domain sockets:
2024-05-13 16:14:27,968  '/run/user/1000/xpra/clients/client-host-298575'
2024-05-13 16:14:28,345 No OpenGL_accelerate module loaded: No module named 'OpenGL_accelerate'
2024-05-13 16:14:28,539 OpenGL enabled on 'AMD Radeon Graphics'
2024-05-13 16:14:28,551  keyboard settings: rules=evdev, model=pc105, layout=us
2024-05-13 16:14:28,553  desktop size is 2560x1440:
2024-05-13 16:14:28,553   :0.0 (677x381 mm - DPI: 96x96) workarea: 2560x1420
2024-05-13 16:14:28,553     SAM HDMI-A-1     (597x336 mm - DPI: 109x109)
2024-05-13 16:14:29,162 enabled remote logging
2024-05-13 16:14:29,163 Xpra X11 seamless server version 6.0
2024-05-13 16:14:29,206 Attached to xpra server at tcp://localhost:12345/6
2024-05-13 16:14:29,206  (press Control-C to detach)
```


Previous Arch `xpra-4.4.5-3` client did not have this issue, although it could never successfully use OpenGL in the first place, at least in my experience.

Setting `--opengl=no` seems to solve the issue, or at least make it unnoticeable.

**Comment by alexisshaw** [comment]
Xpra client 6.0 on Windows has this issue, and the v5.0.8 client doesn't, though it successfully used openly on 5.0.8

**Comment by totaam** [comment]
v6.0.1 will have scroll encoding disabled until I can figure out where the problem comes from: a65e0e24b3022db5679c0deab9f20d8d8f218b64

**Comment by totaam** [comment]
I had forgotten to disable it again for 6.1: 8c05f445d6bb9f220adb3bed3ee4b13a652e3676

After adding support for another fast hashing function: [google's cityhash](https://github.com/google/cityhash) and replacing [xxh3](https://github.com/Cyan4973/xxHash):
```diff
diff --git a/xpra/server/window/motion.pyx b/xpra/server/window/motion.pyx
index 1f7928eff3..b5c2beea17 100644
--- a/xpra/server/window/motion.pyx
+++ b/xpra/server/window/motion.pyx
@@ -15,7 +15,7 @@ from xpra.log import Logger
 log = Logger("encoding", "scroll")
 
 from xpra.buffers.membuf cimport memalign, buffer_context  # pylint: disable=syntax-error
-from xpra.buffers.xxh cimport xxh3
+from xpra.buffers.cityhash cimport cityhash64
 from xpra.util.rectangle import rectangle
 
 
@@ -124,7 +124,7 @@ cdef class ScrollData:
             assert row_len<=rowstride, "invalid row length: %ix%i=%i but rowstride is %i" % (width, bpp, width*bpp, rowstride)
             with nogil:
                 for i in range(height):
-                    a2[i] = xxh3(buf, row_len)
+                    a2[i] = cityhash64(buf, row_len)
                     buf += rowstride
 
 
```

We can also use the much slower `hashlib` variants:
```diff
diff --git a/xpra/server/window/motion.pyx b/xpra/server/window/motion.pyx
index 1f7928eff3..74076eb83d 100644
--- a/xpra/server/window/motion.pyx
+++ b/xpra/server/window/motion.pyx
@@ -117,15 +117,18 @@ cdef class ScrollData:
         cdef uint64_t *a2 = self.a2
         cdef uint16_t i
         cdef uint8_t *buf
+        from hashlib import sha256
+        from struct import unpack
         with buffer_context(pixels) as bc:
             buf = <uint8_t*> (<uintptr_t> int(bc))
             assert len(bc)>=min_buf_len, "buffer length=%i is too small for %ix%i with rowstride %i, should be %i" % (
                     len(bc), width, height, rowstride, min_buf_len)
             assert row_len<=rowstride, "invalid row length: %ix%i=%i but rowstride is %i" % (width, bpp, width*bpp, rowstride)
-            with nogil:
-                for i in range(height):
-                    a2[i] = xxh3(buf, row_len)
-                    buf += rowstride
+            for i in range(height):
+                row = buf[:row_len]
+                digest = sha256(row, usedforsecurity=False).digest()
+                a2[i] = unpack(">Q", digest[:8])[0]
+                buf += rowstride
 
 
     def calculate(self, uint16_t max_distance=1000) -> None:
```
The visual artifacts persists.

So it's unlikely to be caused by unexpected 64-bit hash collisions.

**Comment by totaam** [comment]
References:
* [framebuffers](https://open.gl/framebuffers)
* [Framebuffer Blitting](https://www.khronos.org/opengl/wiki/Framebuffer#Blitting)
* [`glFramebufferTexture2D`](https://registry.khronos.org/OpenGL-Refpages/es2.0/xhtml/glFramebufferTexture2D.xml)
* [`glReadBuffer`](https://registry.khronos.org/OpenGL-Refpages/gl4/html/glReadBuffer.xhtml) / [`glDrawBuffer`](https://registry.khronos.org/OpenGL-Refpages/gl4/html/glDrawBuffer.xhtml)

Should we be using render buffers instead of textures?

**Comment by totaam** [comment]
I can't hold up the 6.1.1 release any longer, so `scroll` is disabled in the client again..

**Comment by totaam** [comment]
TLDR: use `glClear(GL_COLOR_BUFFER_BIT)`.
I don't understand why the fix works, but it does: 617aa2e92bb6b10ed2e8b16c21b4341a77913a93 was a fix for desktop scaling (#4324) and applying the same fix to scroll paints fixes that too: 342dfa0bec7e6d16d3b277364bb0f52b160a209f

---

Why do we need to `glClear` the FBO since we're copying 100% of the source FBO into it immediately after?
Why would the previous contents matter at all?
And the textures attached to the fbos should have been cleared anyway since we `glTexImage2D` them with a NULL buffer on the statement just before that!?:
https://github.com/Xpra-org/xpra/blob/342dfa0bec7e6d16d3b277364bb0f52b160a209f/xpra/opengl/backing.py#L580
